upstairs/src/volume.rs

```diff
@@ -3490,55 +3510,116 @@ mod test {
+        assert!(sv_count > sv_changed);
+        assert!(cid < 3);
         // A valid replacement VCR is provided with a larger generation
         // number and only one target being different.
+        let vol_id = Uuid::new_v4();
+        let opts = generic_crucible_opts(vol_id);
+
         let block_size = 512;
         let vol_id = Uuid::new_v4();
         let blocks_per_extent = 10;
         let extent_count = 9;
 
-        let mut opts = generic_crucible_opts(vol_id);
+        let rop = if add_rop {
+            let rop = Box::new(VolumeConstructionRequest::Region {
+                block_size,
+                blocks_per_extent,
+                extent_count,
+                opts: opts.clone(),
+                gen: 3,
+            });
+            Some(rop)
+        } else {
+            None
+        };
 
-        let original = VolumeConstructionRequest::Volume {
-            id: vol_id,
-            block_size,
-            sub_volumes: vec![VolumeConstructionRequest::Region {
+        let mut sub_volumes = Vec::new();
+        for _ in 0..sv_count {
+            let sv = VolumeConstructionRequest::Region {
                 block_size,
                 blocks_per_extent,
                 extent_count,
                 opts: opts.clone(),
                 gen: 2,
-            }],
-            read_only_parent: None,
+            };
+            sub_volumes.push(sv);
+        }
+
+        let original = VolumeConstructionRequest::Volume {
+            id: vol_id,
+            block_size,
+            sub_volumes,
+            read_only_parent: rop.clone(),
         };
 
         // Change just the minimum things and use the updated values
         // in the replacement volume.
         let original_target = opts.target[cid];
         let new_target: SocketAddr = "127.0.0.1:8888".parse().unwrap();
-        opts.target[cid] = new_target;
 
-        let replacement = VolumeConstructionRequest::Volume {
-            id: vol_id,
-            block_size,
-            sub_volumes: vec![VolumeConstructionRequest::Region {
+        let mut replacement_sub_volumes = Vec::new();
+        for i in 0..sv_count {
+            let mut replacement_opts = opts.clone();
+            let replacement_gen = 3;
+
+            if i == sv_changed {
+                replacement_opts.target[cid] = new_target;
+            }
+
+            let sv = VolumeConstructionRequest::Region {
                 block_size,
                 blocks_per_extent,
                 extent_count,
-                opts: opts.clone(),
-                gen: 3,
-            }],
-            read_only_parent: None,
+                opts: replacement_opts.clone(),
+                gen: replacement_gen,
+            };
+            replacement_sub_volumes.push(sv);
+        }
+        let replacement = VolumeConstructionRequest::Volume {
+            id: vol_id,
+            block_size,
+            sub_volumes: replacement_sub_volumes,
+            read_only_parent: rop.clone(),
```

Comment:
Nit: similarly, this could be
```rust
        let replacement_sub_volumes = (0..sv_count)
            .map(|i| {
                let mut replacement_opts = opts.clone();
                let replacement_gen = 3;

                if i == sv_changed {
                    replacement_opts.target[cid] = new_target;
                }

                VolumeConstructionRequest::Region {
                    block_size,
                    blocks_per_extent,
                    extent_count,
                    opts: replacement_opts.clone(),
                    gen: replacement_gen,
                }
            })
            .collect();
```

Reply:
In 1949ae7d I did this as well as made a few into a common function.